### ISSUE TITLE
Limit I/O throughput for VMs 

### ISSUE THREAD

**Issue by Rot127** [QubesOS/qubes-issues#6660]
**The problem you're addressing (if any)**

If a I/O hungry process is running on a HVM (like Android Studio) the harddrives get sometimes used  so extensively that the systems basically dead locks. The mouse pointer starts lagging or everything freezes.  

**Describe the solution you'd like**

A limit for maximum I/O operations per VM.  is the value to a user, and who might that user be?**

**Describe alternatives you've considered**

**Additional context**

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

**Comment by DemiMarie** [comment]
I am marking this `P: major` because it is a hard requirement for hostile multi-tenancy scenarios, where denial of service is a legitimate security concern.  But I am also marking it as Milestone: TBD because while we do want to support those scenarios someday, there are higher priorities at the moment.

**Comment by andrewdavidwong** [comment]
This sounds similar to https://github.com/QubesOS/qubes-issues/issues/6659. Is it a duplicate?

**Comment by DemiMarie** [comment]
> This sounds similar to #6659. Is it a duplicate?
> 
> To clarify: I realize that I/O and swap are different. They are not duplicates in that sense. But are they both trying to solve the same underlying problem?

No.  #6659 is about a VM swapping too much causing *that VM* to freeze.  This is about *cross-VM* denial of service.

**Comment by Rot127** [comment]
I forgot to mention some important points.
- This only happens if a lot (8-10+) VMs are running.
- The HVM which causes the freezes has a usb-mouse and keyboard attached. It is possible that only the HVM froze and the system could still be used via the PCI-Touchpad, but I didn't recognized it sometimes.
- The little widget in the task bar, which shows the CPU usage, freezes as well. This is just a confirmation that dom0 is affected.
- I do not know how to reproduce the critical throughput of I/O operations, but starting 5< VMs at once and the cursor starts lagging a bit.

**Comment by jamke** [comment]
I have a question that's related to this ticket.
Is it possible to guarantee that one AppVM runs at least with 1 CPU and other resources not matter what happens in other AppVM?

Currently I play 1080p youtube if Firefox in AppVM1 and download torrent in AppVM2. When downloading goes in AppVM2 (to a completely unrelated hdd) the video in AppVM1 freezes and is almost unwatchable. 
Both AppVMs have 1 CPU. I also tried to give AppVM1 all available CPUs (VCPUs no. option).

It's something about sharing some I/O? Because 1 CPU is more than enough to play youtube 1080p in my case.

Any solutions or workarounds will be appreciated.

**Comment by Rot127** [comment]
As of my understanding the VCPU option does not bind a specific hardware CPU (or CPU thread) to the VM, but only a virtual one (VCPU = Virtual CPU).
The instructions executed on the virtual CPU get distributed between the actual hardware CPUs.

Your case sounds like it could also be a network problem. Maybe your download speed is not high enough for download + video?

In this issue we discuss extensive read/write operations to the hard drive and RAM. Which, please correct me if I am wrong, does not seem to be related to your problem.

Jun 5, 2021 11:29:47 jamke ***@***.***>:

**Comment by jamke** [comment]
> Your case sounds like it could also be a network problem. Maybe your download speed is not high enough for download + video?

Thanks for reply. I'm sure it's not about with network speed as I can get the same situation with offline AppVM2. It's either CPU, RAM or some I/O issues in Qubes OS.
Now I think it's probably CPU power distribution. So, in Xen/QubesOS there is no way to guarantee some minimal CPU power for particular AppVM, am I right?

> In this issue we discuss extensive read/write operations to the hard drive and RAM. Which, please correct me if I am wrong, does not seem to be related to your problem.

Yes, maybe it's not related, I can't tell.
In any case if anyone ever a archived isolation of resources for uninterruptible playback of video in AppVM in Qubes OS - please let me know somehow.

**Comment by Rot127** [comment]
Just found out a great way to simulate this problem. Simply restore a VM from a backup or install a template while working on the PC.
The system will freeze occasionally.

Technically this is not the same problem as described in the issue (dom0 is the source of the huge resource usage, not some unprivileged VM), but I mention it because it produces the same symptoms and is maybe related.

Besides. Should dom0 be allowed to use that much resources so the system freezes?
